### PR TITLE
Added plugin to to decorate output files with `NS_ASSUME_NONNULL_*` macros.

### PR DESCRIPTION
**Pull request by GotEmB**
This PR adds a plugin, which when required will decorate your value object's interface and implementation with the `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` macros.

### CODE REVIEW

File: src/__tests__/plugins/assume-nonnull-tests.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        typeLookups:[],
+        excludes: [],
+        includes: ['AssumeNonnull'],
+        libraryName: Maybe.Nothing<string>(),
+        typeName: 'Foo'
+      };
+
+      const imports:ObjC.Import[] = ValueObjectPlugin.imports(valueType);
+      const expectedImports:ObjC.Import[] = [
+        { file: 'Foundation2.h', isPublic: true, library: Maybe.Just('Foundation2') },
```
Comment (ColinCampbell) [inline src/__tests__/plugins/assume-nonnull-tests.ts]: This doesn't seem right? You're also not implementing imports in your plugin so I don't think we need this test

File: src/plugins/assume-nonnull.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+import AlgebraicType = require('../algebraic-type');
+import Code = require('../code');
+import Error = require('../error');
+import FileWriter = require('../file-writer');
+import Maybe = require('../maybe');
+import ObjC = require('../objc');
+import ValueObject = require('../value-object');
+
+function assumeNonnullFileTransformation(request:FileWriter.Request):FileWriter.Request {
```
Comment (ColinCampbell) [inline src/plugins/assume-nonnull.ts]: We shouldn't be using a file transformation for this. It's going to make the plugin more brittle and make it more likely that it won't be compatible with other plugins which don't output things in the same way.

This is going to be more work, but I would recommend

- Adding a `nullability` property to `Class` in `objc.ts`
- It should be an enum type `ObjC.ClassNullability` that has something like `default` and `assumeNonnull`
- You can then add a function to the value and algebraic type plugins which can return a Maybe.Maybe<ObjC.ClassNullability>
- Your plugin can just `Maybe.Just(ObjC.ClassNullability.assumeNonnull`
- Change `objc-renderer.ts` to output the nullability macros for classes with the `assumeNonnull` value for `nullability`

Comment (GotEmB) [inline src/plugins/assume-nonnull.ts]: @ColinCampbell Yeah what you're suggesting seems like a better approach. It should also make writing unit tests better compared to what I have right now.

File: src/objc-renderer.ts [hunk 3]
```diff
@@ -243,14 +243,77 @@ function enumerationIsPublic(isPublic:boolean):(enumeration:ObjC.Enumeration) =>
...
 }
 
+const NS_ASSUME_NONNULL_BEGIN:string = 'NS_ASSUME_NONNULL_BEGIN';
+const NS_ASSUME_NONNULL_END:string = 'NS_ASSUME_NONNULL_END';
+
+function prefixClassMacros(classInfo:ObjC.Class):string[] {
+  var macros:string[] = [];
+  switch (classInfo.nullability) {
+    case ObjC.ClassNullability.assumeNonnull:
+      macros.push(NS_ASSUME_NONNULL_BEGIN);
```
Comment (ColinCampbell) [inline src/objc-renderer.ts]: We shouldn't be mutating the array from inside the if. For now, I would simply return the array from here. In the future when we add more, we can figure out the right way to build the array up

File: src/objc-renderer.ts [hunk 4]
```diff
@@ -243,14 +243,77 @@ function enumerationIsPublic(isPublic:boolean):(enumeration:ObjC.Enumeration) =>
...
+      break;
+  }
+  return macros;
+}
+
+function postfixClassMacros(classInfo:ObjC.Class):string[] {
+  var macros:string[] = [];
+  switch (classInfo.nullability) {
+    case ObjC.ClassNullability.assumeNonnull:
+      macros.push(NS_ASSUME_NONNULL_END);
```
Comment (ColinCampbell) [inline src/objc-renderer.ts]: Samesies

File: src/objc-renderer.ts [hunk 5]
```diff
@@ -243,14 +243,77 @@ function enumerationIsPublic(isPublic:boolean):(enumeration:ObjC.Enumeration) =>
...
+      break;
+  }
+  return macros;
+}
+
+function prefixBlockTypeMacros(blockType:ObjC.BlockType):string[] {
+  var macros:string[] = [];
+  switch (blockType.nullability) {
+    case ObjC.ClassNullability.assumeNonnull:
+      macros.push(NS_ASSUME_NONNULL_BEGIN);
```
Comment (ColinCampbell) [inline src/objc-renderer.ts]: Samesies

File: src/objc-renderer.ts [hunk 6]
```diff
@@ -243,14 +243,77 @@ function enumerationIsPublic(isPublic:boolean):(enumeration:ObjC.Enumeration) =>
...
+      break;
+  }
+  return macros;
+}
+
+function postfixBlockTypeMacros(blockType:ObjC.BlockType):string[] {
+  var macros:string[] = [];
+  switch (blockType.nullability) {
+    case ObjC.ClassNullability.assumeNonnull:
+      macros.push(NS_ASSUME_NONNULL_END);
```
Comment (ColinCampbell) [inline src/objc-renderer.ts]: Samesies

Comment (ColinCampbell) [inline src/objc-renderer.ts]: Something to consider is that we're also switching on this enum in four separate places. Did you think about pulling out a MacroLocations type which has four different properties (`classPrefix`, `classPostfix`, etc) or maybe even a Macro type with `prefix` and `postfix` and have MacroLocations contain two Macro types for `class` and `block`

File: features/assume-nonnull.feature [hunk 7]
```diff
@@ -0,0 +1,236 @@
+Feature: Outputting Value Objects / Algebraic Types decorated with NS_ASSUME_NONNULL_* macros
+
+  @announce
+  Scenario: Generation header and method files for Value Objects with the NS_ASSUME_NONNULL_* macros
+    Given a file named "project/values/RMFoo.value" with:
+      """
+      RMFoo includes(RMAssumeNonnull) {
+        NSString *aString;
+        %nullable NSString *bString;
```
Comment (GotEmB) [inline features/assume-nonnull.feature]: Haha really? I thought this was the syntax 😂

File: src/objc-renderer.ts (hunk 6 above)
Comment (GotEmB) [inline src/objc-renderer.ts]: I kinda thought about that, but then saw that most of everything else was structured this way. But yeah refactoring this into a `Macro` type and a `MacroLocation` type would make this cleaner. I'll refactor them.

Comment (ColinCampbell) [inline src/objc-renderer.ts]: Just a reminder on this :)

Comment (GotEmB) [inline src/objc-renderer.ts]: @ColinCampbell, sorry for the delay. I'll try to get to this today / tomorrow.

**Pull request merged**